Jenkinsfile

```diff
@@ -0,0 +1,91 @@
+#!groovy
+
+/*
+ * Copyright © 2017 IBM Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+
+stage('Build') {
+    // Checkout, build and assemble the source and doc
+    node {
+        checkout scm
+        sh './gradlew clean assemble'
+        stash name: 'built'
+    }
+}
+
+stage('Test') {
+    node {
+        // Unstash the built code
+        unstash name: 'built'
+        withCredentials([[$class: 'UsernamePasswordMultiBinding', credentialsId: 'clientlibs-test', usernameVariable: 'DB_USER', passwordVariable: 'DB_PASSWORD']]) {
+            // Run the core and in-process tests
+            // Note Redis server not available in this env, so those tests excluded (but tested by Travis CI)
+            try {
+                sh "./gradlew -Dtest.couch.url=https://${env.DB_USER}:${env.DB_PASSWORD}@clientlibs-test.cloudant.com :cloudant-client-cache:test :cloudant-client-cache-in-process:test"
+            } finally {
+                junit '**/build/test-results/*.xml'
+            }
+        }
+    }
+}
+
+// Publish the master branch
+stage('Publish') {
+    if (env.BRANCH_NAME == "master") {
+        node {
+            checkout scm // re-checkout to be able to git tag
+            unstash name: 'built'
+            // read the version name and determine if it is a release build
+            version = readFile('VERSION').trim()
+            isReleaseVersion = !version.toUpperCase(Locale.ENGLISH).contains("SNAPSHOT")
+
+            // Upload using the ossrh creds (upload destination logic is in build.gradle)
+            withCredentials([usernamePassword(credentialsId: 'ossrh-creds', passwordVariable: 'OSSRH_PASSWORD', usernameVariable: 'OSSRH_USER'), usernamePassword(credentialsId: 'signing-creds', passwordVariable: 'KEY_PASSWORD', usernameVariable: 'KEY_ID'), file(credentialsId: 'signing-key', variable: 'SIGNING_FILE')]) {
+                sh './gradlew -Dsigning.keyId=$KEY_ID -Dsigning.password=$KEY_PASSWORD -Dsigning.secretKeyRingFile=$SIGNING_FILE -DossrhUsername=$OSSRH_USER -DossrhPassword=$OSSRH_PASSWORD upload'
+            }
+
+            // if it is a release build then do the git tagging
+            if (isReleaseVersion) {
+
+                // Read the CHANGES.md to get the tag message
```

Comment:
Looking at this again I recalled that my original implementation of this (in whichever repo was first) did follow the more concise groovy file reading pattern. However, that didn't work in the `Jenkinsfile` because I had to use `readFile` to get access to the workspace content. That method doesn't return contents in an iterable way (it's just a string). So the `""" """` again serves to coerce the string into a multiline string. I think we could probably use `''' '''` since I don't think we need to do any interpolation and just need multi-line to get access to the `readLines` method, but I think I felt it was more confusing to use two different string types in the same block.
There is an alternative to use some Java classes to read the string in lines, but each method needs in-process script approval, so it was easier to stick with standard methods available from the groovy for the `Jenkinsfile`.
Fun right?